### ISSUE TITLE
split apart models.py?

### ISSUE THREAD

**Issue by binary1230** [magfest/ubersystem#1519]
Hey, models.py is getting really huge these days.

Are there any strong reasons we can't split it out into a bunch of files like models/attendee.py models/admin_account.py /etc?

I have purely selfish motivations here, as hopping around in this file is occasionally confusing to remember what part of it you're in.

**Comment by EliAndrewC** [comment]
> Hey, models.py is getting really huge these days.

Funny you should say that, since ``models.py`` is several hundred lines shorter than it was before we started pulling things out into plugins.

Your point is well taken though.  I personally prefer having everything in one big file there, an until it gets above a couple of thousand lines of code (we're a bit under 2kloc now) then I'm fine with it.

However, if we wanted to, we could easily do what you're suggesting, though I'd be strongly against doing a single file per class, since I think that's WAY too much splitting up.  The way I'd organize it if we split it up would be:
- ``uber/models/__init__.py`` -> mostly empty file, the only thing it does is import the sub-modules, probably in the following order
- ``uber/models/utils.py`` -> all of our assorted utility functions and classes, e.g. ``Choice``, ``MultiChoice``, ``utcnow``, our custom modifications to some of the SQLAlchemy builtins like ``relationship``, etc
- ``uber/models/base.py`` -> our ``MagModel`` base class
- ``uber/models/session.py`` -> our ``Session`` class
- ``uber/models/schema.py`` -> all of our ``MagModel`` subclasses

If we split it up that way, none of the files above would break a thousand lines of code.  My personal preference would be to keep it all together, but that's a fairly mild preference, so I wouldn't object to the above.

**Comment by kitsuta** [comment]
As someone who's very familiar with models.py, I like Eli's structure.. but I worry that it may be a little too obtuse for someone new.

**Comment by thaeli** [comment]
I would strongly prefer to keep this all together for now. It's much easier to have one place to look than a bunch IMO.

**Comment by binary1230** [comment]
Yea my reasons are pretty superficial, I was finding it easy to get lost when jumping around between different models and utility functions in here.  Specifically in my case I was playing around with insert_test_admin() and having to scroll around in this file a lot.

No need to change anything right now, though I might advocate soon for splitting Session into it's own thing.

**Comment by jmdawson** [comment]
I would solve this problem in my familiar language with a package full of small files.  You have a _lot_ of classes going on in that file, and it's inscrutable.  Judging by the size, I'd say it's actually several packages worth of things.  That also mostly solves the 'one place to look for things,' because you look in the package.  Navigation in that way also means you can get more information out of the package structure, and so new people find it easier to contribute.

tl;dr: A 2000-line file with _many_ classes in it is not easy for a new person to contribute to, and I hate it because I'm a java developer.

**Comment by kitsuta** [comment]
So, uh, what is a package in this context?

**Comment by EliAndrewC** [comment]
What @jmdawson suggested would basically be what I proposed in my original comment, except that it would be a single file per class, which is how Java is usually written.  For the reasons I explained in my original comment I'd be against that level of splitting up (and honestly I prefer things how they are now, but I don't have a super-strong preference).

**Comment by jmdawson** [comment]
yeah, a package in this sense would be a folder basically called models.  Or domain.  Or several folders since it's hard to tell what's going on exactly in this file, and also since there are general guidelines that you shouldn't have _too_ many classes in a single package, because it becomes hard to figure out what is going on in there.  But I'd start w/ one file per class, or possibly tolerate a few highly related classes in one file (like the two MAGPoints classes) as a concession to the weird style you seem to have adopted.

Really, aside from personal preference, what is the reasoning behind NOT divvying them up so you have one file per class?

**Comment by EliAndrewC** [comment]
> Really, aside from personal preference, what is the reasoning behind NOT divvying them up so you have one file per class?

It's really just "personal preference" honestly.  But here are the reasons for my preference:
* I do think it's easier to find things.  "This is where all of the database tables and their models are defined" makes it really easy to know what file to look in.
* When reviewing code for e.g. jobs/shifts it's more convenient to scroll up and down than to switch back and forth between tabs.
* Relatedly, we have about two dozen classes in that file, so having to open two dozen tabs is much more annoying.
* Even if we split it up, the reason why I'd prefer to group related classes together is that I think it'll be much more obvious where to look.  For example, someone might open the ``Shift`` file instead of the ``Job`` file because it's not really apparent which one they want.  (Regardless we ought to have better documentation, which we've gradually been backfilling.)

None of these are slam dunk reasons, which is why as I said, I don't have a particularly strong opinion on this.

**Comment by kitsuta** [comment]
Yes, I definitely want someone to be able to figure out where things are without memorizing it, mostly because I have a terrible memory.

A single file makes this doable, sorta, because everything is in one place. But I've frequently felt the pain of:
* Accidentally editing/looking at e.g. a Group class property when I wanted to edit an Attendee property of the same name, because I'm using ctrl-F to navigate the document.
* Constantly having to bounce back and forth between two disparate areas of the file, e.g. a SessionMixin class and the Attendee class function that calls it.

I definitely see Eli's point, but the fact is that in most cases you won't *open* two dozen classes. You'll open three or four, max.

We could do this in one of two ways:
* Folders of functionally-grouped class files, e.g. a "Base" folder with MagModel, Session, and SessionMixin class files and then a "Registration" folder with Group and Attendee class files.
* One folder for all DB stuff, then files grouped as if they were the folders above. So base.py, registration.py, etc.

I actually slightly prefer the former, despite the fact that I don't love navigating through a bunch of folders, because it means that if you're looking for "the Attendee class" you will be able find attendee.py and you know what the heck that contains without opening it. The latter carries the risk of the developer having to open five different files just to find the exact class they want.

**Comment by thaeli** [comment]
+1 for "leave models.py monolithic". Honestly the one tiny class per file thing is one of the reasons I dread working on Java code so much. Personal preference, yes, and also I'd point out that plugin specific models are ultimately the place a bunch of this has been or will be refactored out to anyway.

**Comment by jmdawson** [comment]
The current models file is completely inscrutable from my perspective.  A move towards separating unrelated classes would be a major step forward.

The thing is, your class files in Java should be cohesive and loosely coupled, so you know that what is in there does _one_ thing.  You don't need to look at it all that often, which means you can set it aside and focus on what ever else it is you're doing.  It also reduces the needle-in-haystack issue.  _Plus_ it means that you don't end up with methods potentially scattered all over a monolithic file.

Monoliths are for monument parks, not code.

**Comment by binary1230** [comment]
wouldn't mind one thing which people are suggesting which is: split out anything that isn't derived from MagModel into another file.

There is one other also slightly superficial reason to split the files: In the event of a larger refactor where changes happen to lots of classes in models.py in different files, it might be easier to re-merge those changes. though..... in that scenario it's a mess either way so not sure.

**Comment by EliAndrewC** [comment]
> Monoliths are for monument parks, not code.

https://www.youtube.com/watch?v=pWdd6_ZxX8c

:P

**Comment by RobRuana** [comment]
Done and done!
